[fix] 프로덕션 JVM Metaspace 상한 부족으로 OOM 발생 및 좀비 상태 지속

## 버그 설명

프로덕션 백엔드 컨테이너가 2026-04-18 22:36:33 KST부터 `java.lang.OutOfMemoryError: Metaspace`를 반복 발생시키며 502 Bad Gateway를 반환. 약 3시간 동안 서비스 장애. 수동 `docker restart`로 복구.

핵심 원인은 두 가지:

1. `.env.prod`의 `JAVA_OPTS`에 설정된 `-XX:MaxMetaspaceSize=128m`이 Spring Boot 3.5 + Security + OAuth2 + Springdoc + JPA + Flyway + Actuator 조합의 베이스라인(약 110~120MB)에 비해 안전 마진이 없음. OAuth2 첫 Google 로그인 시점에 TLS 관련 JDK 클래스(`jdk.crypto.ec/EdDSAParameters`, `sun.security.ssl.SignatureScheme`)가 추가 로딩되며 128m 천장을 넘김.
2. Metaspace OOM 시 JVM 기본 동작은 프로세스를 종료하지 않고 요청마다 OOM을 뱉음. `restart: unless-stopped`가 있어도 JVM이 죽지 않으니 docker가 재시작하지 못해 좀비 상태가 지속됨(헬스체크 779회 연속 실패).

## 재현 방법

1. `MaxMetaspaceSize=128m` 상태로 Spring Boot 백엔드를 장시간(수십 시간) 운영
2. 운영 중 신규 Google OAuth2 로그인이 처음 발생하여 TLS 시그니처 스키마 클래스가 로딩되는 시점 도달
3. Metaspace가 128m를 초과하며 `OutOfMemoryError: Metaspace` 발생
4. JVM은 종료되지 않고 모든 후속 요청에 OOM 반환. Nginx는 unhealthy upstream에 502 응답

> 저트래픽 환경에서 자연 발생한 이슈라 로컬 재현은 `-XX:MaxMetaspaceSize=80m` 수준으로 낮춰 기동하면 유사 조건 재현 가능.

## 예상 동작

- Metaspace 상한이 현재 의존성 기준으로 충분한 여유를 가질 것 (256m 이상)
- OOM이 발생하더라도 JVM이 즉시 종료되어 docker `restart: unless-stopped`에 의해 자동 복구될 것
- OOM 발생 시 heap dump가 남아 후속 진단 가능할 것

## 환경

- OS: Ubuntu (프로덕션 서버)
- JDK: Temurin 21.0.10
- Spring Boot: 3.5.11
- 컨테이너: `ghcr.io/rabemaster/gakkaweo-backend:latest` (Docker Compose)
- 발생 시각: 2026-04-18 22:36:33 KST
- 장애 지속: 약 3시간 (수동 재시작까지)

## 참고자료

**수정 대상**: `.env.prod.sample` (커밋 대상) + 서버의 `.env.prod` (수동 동기화)

```diff
-JAVA_OPTS=-Xms256m -Xmx512m -XX:+UseG1GC -XX:MaxMetaspaceSize=128m
+JAVA_OPTS=-Xms256m -Xmx512m -XX:+UseG1GC -XX:MaxMetaspaceSize=256m -XX:+ExitOnOutOfMemoryError -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=/tmp/heapdump.hprof
```

**배포 절차**

1. PR 머지 후 서버 `.env.prod`의 `JAVA_OPTS` 라인을 sample과 동일하게 수동 동기화
2. `docker compose -f docker-compose.prod.yml up -d --force-recreate backend`로 env 재적용

**OOM 시 로그 스택 발췌** (docker logs 기준)
```
Caused by: java.lang.OutOfMemoryError: Metaspace
  at sun.security.util.math.intpoly.IntegerPolynomial.getElement
  at jdk.crypto.ec/sun.security.ec.ed.EdDSAParameters.<clinit>
  at jdk.crypto.ec/sun.security.ec.ed.EdDSASignature.<init>
  at java.security.Signature.getInstance
  at sun.security.ssl.SignatureScheme.<init>
```

--- a/.env.prod.sample
+++ b/.env.prod.sample
@@ -60,6 +60,7 @@ LOG_LEVEL=INFO
 
 # JVM
 # MaxMetaspaceSize: Spring Boot 3.5 + Security + OAuth2 + Springdoc 베이스라인 기준 256m 이상 필요
-# ExitOnOutOfMemoryError: OOM 발생 시 JVM 즉시 종료 → docker restart 정책이 자동 복구 (좀비 상태 방지)
-# HeapDumpOnOutOfMemoryError: OOM 진단용 힙 덤프 생성 (컨테이너 재시작 시 소실됨, 사후 분석 시 docker cp로 회수)
-JAVA_OPTS=-Xms256m -Xmx512m -XX:+UseG1GC -XX:MaxMetaspaceSize=256m -XX:+ExitOnOutOfMemoryError -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=/tmp/heapdump.hprof
+# ExitOnOutOfMemoryError: OOM 발생 시 JVM 즉시 종료 -> docker restart 정책(unless-stopped)이 자동 복구 (좀비 상태 방지)
+# HeapDumpOnOutOfMemoryError: OOM 진단용 힙 덤프 생성
+# HeapDumpPath: /data/dumps는 docker-compose.prod.yml에서 호스트의 ./dumps에 마운트된 전용 볼륨이라 재시작 후에도 보존
+JAVA_OPTS=-Xms256m -Xmx512m -XX:+UseG1GC -XX:MaxMetaspaceSize=256m -XX:+ExitOnOutOfMemoryError -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=/data/dumps/heapdump.hprof

--- a/.env.prod.sample
+++ b/.env.prod.sample
@@ -59,4 +59,7 @@ FORWARD_HEADERS_STRATEGY=native
 LOG_LEVEL=INFO
 
 # JVM
-JAVA_OPTS=-Xms256m -Xmx512m -XX:+UseG1GC -XX:MaxMetaspaceSize=128m
+# MaxMetaspaceSize: Spring Boot 3.5 + Security + OAuth2 + Springdoc 베이스라인 기준 256m 이상 필요
+# ExitOnOutOfMemoryError: OOM 발생 시 JVM 즉시 종료 → docker restart 정책이 자동 복구 (좀비 상태 방지)
+# HeapDumpOnOutOfMemoryError: OOM 진단용 힙 덤프 생성 (컨테이너 재시작 시 소실됨, 사후 분석 시 docker cp로 회수)
+JAVA_OPTS=-Xms256m -Xmx512m -XX:+UseG1GC -XX:MaxMetaspaceSize=256m -XX:+ExitOnOutOfMemoryError -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=/tmp/heapdump.hprof

--- a/docker-compose.prod.yml
+++ b/docker-compose.prod.yml
@@ -74,6 +74,7 @@ services:
       - "127.0.0.1:${SERVER_PORT}:8080"
     volumes:
       - ./uploads:/data/uploads
+      - ./dumps:/data/dumps
     env_file: .env.prod
     networks:
       - internal

